instrumentation/CONTRIBUTING.md

```diff
@@ -0,0 +1,407 @@
+
+For example, the `Werewolf` generated in the example above instrumentation is available `OpenTelemetry::Instrumentation::Werewolf::Instrumentation.instance.tracer`. You should reference this tracer in your code when creating spans like this:
+
+```ruby
+
+  OpenTelemetry::Instrumentation::Werewolf::Instrumentation.instance.tracer.start_span('transform') do
+    # code to be traced
+  end
+
+```
+
+> :warning: This tracer is not _upgradable_ before the SDK is initialized, therefore it is important that your instrumentation _always_ use stack local references of the tracer.
+
+### Use First Party Extension Points
+
+Whenever possible, use first party extension points (hooks) to instrument libraries. This ensures that the instrumentation is compatible with the latest versions of the library and that the instrumentation is maintained by the library authors. `ActiveSupport::Notifications` and `Middleware` are good examples of a first party extension points that are used by our instrumentation libraries.
+
+Monkey patching is discouraged in OpenTelemetry Ruby because it is the most common source of bugs and incompatability with the libraries we are instrumenting. If you must monkey patch, please ensure that the monkey patch is as isolated as possible and that it is clearly documented.
+
+### Use Semantic Conventions
+
+Use the OpenTelemetry Semantic Conventions to ensure that the instrumentation is compatible with other OpenTelemetry libraries and that the data is useful in a distributed context.
+
+> :information_source: Privacy and security are important considerations when adding attributes to spans. Please ensure that you are not adding sensitive information to spans. If you are unsure, please ask for a review.
+
+When semantic conventions do not exist, use the [Elastic Common Schema](https://www.elastic.co/guide/en/ecs/current/index.html) and submit a Issue/PR with your attributes to the [Semantic Conventions repo](https://github.com/open-telemetry/semantic-conventions) to propose a new set of standard attributes.
+
+If the attribute is specific to your instrumentation, then consider namespacing it using the `instrumentation` prefix e.g. `werewolf.bite.damage`.
+
+### Write Comprehensive Automated Tests
+
+Code that is not tested will not be accepted by maintainers. We understand that providing 100% test coverage is not always possible but we still ask that you provide your best effort when writing automated tests.
+
+Most of the libraries we are instrumenting introduce changes that are outside of our control. For this reason integration or state based tests are preferred over interaction (mock) tests.
+
+When you do in fact run into cases where test doubles or API stubs are absolutely necessary then we recommend using the `rspec-mocks` and `webmocks` gems.
+
+### Understand Performance Characteristics
+
+Instrumentation libraries should be as lightweight as possible
+
+* Avoid allocating too many objects and private methods
```

Comment:
I think "private methods" should be listed as a separate bullet point, explaining why instrumenting private methods is less desirable.

Reply:
Hmm, actually, because this section is called "Understand Performance Characteristics", what do "private methods" have to do with performance?

Reply:
I need to expand on this a lot!

We want to avoid allocating anything unnecessarily to the call stack, bust the call cache, require our users to add additional backtrace filters, or cause issues with yjit having to unravel code it optimized away as a result of something like an exception. 

I see your point though about private methods seeming to be unrelated and should probably get a section on its own. 

That also brings up another topic/concern with monkey patches and using private methods!

Reply:
I broke out using private methods into it's own section 

https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/946/commits/7285a7d8226086503e23f1f6b0d14ed79b2d5628